src/coreclr/vm/jitinterface.cpp

```diff
@@ -12131,6 +12131,26 @@ HRESULT CEEJitInfo::allocPgoInstrumentationBySchema(
         codeSize = m_ILHeader->GetCodeSize();
     }
 
+    if (m_pMethodBeingCompiled->IsVersionable())
+    {
+        CodeVersionManager* pCodeVersionManager = m_pMethodBeingCompiled->GetCodeVersionManager();
+        CodeVersionManager::LockHolder codeVersioningLockHolder;
+        ILCodeVersion ilVersion = pCodeVersionManager->GetActiveILCodeVersion(m_pMethodBeingCompiled);
+        NativeCodeVersion currentVersion = ilVersion.GetActiveNativeCodeVersion(m_pMethodBeingCompiled);
+
+        NativeCodeVersion::OptimizationTier currentOptTier = currentVersion.GetOptimizationTier();
+        if ((currentOptTier == NativeCodeVersion::OptimizationTier::OptimizationTier0) ||
+            (currentOptTier == NativeCodeVersion::OptimizationTier::OptimizationTier1))
+        {
+            // Current tier is not marked as instrumented, but it requested the schemas so it means it doesn't
+            // need to be re-instrumented in future and can be promoted straight to Tier1 if hot enough.
+            currentVersion.SetShouldSkipInstrumentation();
+        }
+
+        // Leave a note that this method is properly instrumented
+        currentVersion.SetInstrumented();
```

Comment:
`NativeCodeVersion` is a simple wrapper around either a `MethodDesc` or a `NativeCodeVersionNode`, and is typically passed by value. The default native code version doesn't have a node and just wraps a `MethodDesc`. So, it's not a good candidate for persistent storage of this type of info. This change to the code version would just modify the local copy and would not be persisted anywhere.

AFAIK there isn't a clear place currently where such info could be stored. `CallCountingInfo` may be a candidate, for instance see `CallCountingManager::DisableCallCounting()` and `CallCountingInfo::CreateWithCallCountingDisabled()`, which are used to indicate that a code version should not be tiered. However, currently it's not set up to store extra info aside from call counting, for example `CallCountingManager::SetCodeEntryPoint()` does something different if the `CallCountingInfo` already exists as opposed to when it's being created. It may be possible to enable precreating a `CallCountingInfo` for a code version by introducing a new `CallCountingInfo::Stage` something like `NotStarted` to indicate that call counting has not started, and to modify `SetCodeEntryPoint()` to do something similar to before in that case. If the `CallCountingInfo` already exists, it could be used as persistent storage, as it has a hash table mapping code version to a `CallCountingInfo`.

Reply:
Maybe it's better to rollback to occupying a spare bit in `MethodDesc` m_flag2 then? This change is a sort of an optimization, not correctness. The problem with CallCountingStubs that they're created assyncronlsy like you pointed out

Reply:
The info appears to be specific to a code version rather than a method desc, so I'm not sure the method desc would be an appropriate place for it